File: day2/goldStar.py
def is_safe(report):
    if len(report) < 2:
        return True
    diffs = [report[i+1] - report[i] for i in range(len(report)-1)]
    if not all(1 <= abs(d) <= 3 for d in diffs):
        return False
    if all(d > 0 for d in diffs) or all(d < 0 for d in diffs):
        return True
    return False

safe_count = 0

with open('day2/input.txt', 'r') as file:
    for line in file:
        report = list(map(int, line.strip().split()))
        if not report:
            continue
        if is_safe(report):
            safe_count += 1
        else:
            # Try removing each level
            for i in range(len(report)):
                modified_report = report[:i] + report[i+1:]
                if is_safe(modified_report):
                    safe_count += 1
                    break

print(safe_count)

File: R/day2/goldStar.R
# The function purpose is to check if a report is safe
is_safe <- function(report) {
  differences <- diff(report)
  all_increasing <- all(differences > 0 & differences <= 3)
  all_decreasing <- all(differences < 0 & differences >= -3)
  return(all_increasing || all_decreasing)
}

# The function purpose is to check if a report is safe in addition
# with the Problem Dampener
is_safe_with_dampener <- function(report) {
  n <- length(report)
  if (is_safe(report)) {
    return(TRUE)
  }
  for (i in 1:n) {
    modified_report <- report[-i]
    if (is_safe(modified_report)) {
      return(TRUE)
    }
  }
  return(FALSE)
}

data <- readLines("day2/input.txt")

safe_count <- 0

# Process
for (line in data) {
  report <- as.numeric(unlist(strsplit(line, " ")))
  if (is_safe_with_dampener(report)) {
    safe_count <- safe_count + 1
  }
}

cat(safe_count, "\n")